Pegasus/include/pegasus/Primitives.hpp

```diff
@@ -0,0 +1,373 @@
+};
+
+/**
+ * @brief Stores force data
+ *
+ * @attention This class works as a high level API over the Scene singleton instance.
+ * Only the handles are guaranteed to be valid. Any reference/pointer is invalidated
+ * after a new instance of a Force class is created or deleted.
+ *
+ * @tparam ForceType force type
+ */
+template < typename ForceType >
+class Force final : public ForceBase
+{
+public:
+    /**
+     * @brief Allocates new force in the scene and initializes it
+     * @param scene scene instance
+     * @param force initialized force data
+     */
+    Force(Scene& scene, ForceType force)
+        : ForceBase(scene)
+        , m_force(force)
+    {
+        m_forceHandle = m_pScene->MakeForce<ForceType>();
+        SetForce(force);
+    }
+
+    /**
+     * @brief Unbinds all bodies and releases force handle
+     */
+    virtual ~Force()
+    {
+        for (std::pair<Handle, Handle> const bind : m_bodyForceBinds)
+        {
+            m_pScene->UnbindForce<ForceType>(bind.second);
+        }
+        m_pScene->RemoveForce<ForceType>(m_forceHandle);
+    }
+
+    /**
+     * @brief Sets force data
+     * @param force initialized force data
+     */
+    void SetForce(ForceType force)
+    {
+        m_pScene->GetForce<ForceType>(m_forceHandle) = force;
+    }
+
+    /**
+     * @brief Returns force data
+     *
+     * @attention This class works as a high level API over the Scene singleton instance.
+     * Only the handles are guaranteed to be valid. Any reference/pointer is invalidated
+     * after a new instance of a Force class is created or deleted.
+     *
+     * @return force data reference
+     */
+    ForceType& GetForce()
+    {
+        return m_pScene->GetForce<ForceType>(m_forceHandle);
+    }
+
+    /**
+     * @brief Bind force to the given body
+     * @param primitive body to be bound
+     */
+    void Bind(Primitive const& primitive) override
+    {
+        Handle const body = primitive.GetBodyHandle();
+
+        if (m_bodyForceBinds.find(body) == m_bodyForceBinds.end())
+        {
+            m_bodyForceBinds[body] = m_pScene->BindForce<ForceType>(body, m_forceHandle);
+        }
+    }
+
+    /**
+     * @brief Unbinds force from the body and releases bind handle
+     * @param primitive body to be unbound
+     */
+    void Unbind(Primitive const& primitive) override
+    {
+        Handle const body = primitive.GetBodyHandle();
+
+        if (m_bodyForceBinds.find(body) != m_bodyForceBinds.end())
+        {
+            m_pScene->UnbindForce<ForceType>(m_bodyForceBinds[body]);
+            m_bodyForceBinds.erase(body);
+        }
+    }
+
+private:
+    std::unordered_map<Handle, Handle> m_bodyForceBinds;
+    Handle m_forceHandle;
+    ForceType m_force;
+};
+
+/**
+ * @brief Allows multiple similar force bindings
+ */
+class PrimitiveGroup
+{
+public:
+    /**
+     * @brief Adds force pointer to the group and binds the force to all group's bodies
```

Comment:
By using this interface I'm forcing user code to not pass nullptr pointers. I think it's better. And we had this same talk a year ago, and you though so too.

Reply:
I remember it differently.
Problem with passing a reference is ability to pass a reference to stack variable.
If you're explicitly using a pointer then it's on user to make sure that pointer they pass is alive in scope of context of this pointer (if passed to `Bind()` it should be alive up until `Unbind()`). I would suggest using references only when they are used as references or their usage is limited to the scope of the method. If it has to be stored and is used beyond the method body, it shall be passed as pointer.

Reply:
I agree that this is a problem but a see how it could invalidate both ways. And although yes, it is possible and a bit simpler to pass as a reference an object from the stack. It is possible to pass a pointer to the stack variable as well and it would only take one additional operator call. 
My solution here is to use reference as to avoid nullptrs and point out in the interface documentation that the time of life of the variable should be at least as long as the time of life of the Primitive Group object.

Reply:
Although even this is clear from the interface since the only way to unbind is to pass a refence to the same object. So I'm not sure if that makes sence to write that.

Reply:
That makes sense I agree. But again, the interface contains both call to create and to delete, and the only way to delete the bind is to pass the reference to the same object that was used during the creation. I feel like this change would not really gain that much because, on the one hand, I can see how the pointer could be used to indicate the time of life of the object, but on the other I can also see that there is an interface that implies that and also prevents user from passing null pointers.